File: packages/testing/benchmarks/bench-broad-2m.ts
#!/usr/bin/env -S deno run -A --no-check
/**
 * Broad tidy-ts benchmark at 2M rows.
 * Covers: creation, filter, select, sort, mutate, distinct, groupBy,
 *         summarise, innerJoin, leftJoin, pivotLonger, bindRows, stats
 *
 * Prints a 3-way comparison table (tidy-ts vs Polars vs pandas).
 * Re-run bench-broad-2m.py to update the reference numbers below.
 */
import { createDataFrame, stats } from "@tidy-ts/dataframe";

const N = 2_000_000;
const ITERS = 10;
const WARMUP = 5;

// Reference medians from bench-broad-2m.py (2M rows, 10 iters, 5 warmup)
// Re-run: python3 packages/testing/benchmarks/bench-broad-2m.py
const POLARS: Record<string, number> = {
  "creation": 55.883,
  "filter (numeric)": 1.728,
  "filter (string)": 1.326,
  "filter (complex)": 1.489,
  "select": 0.037,
  "sort (numeric)": 72.893,
  "sort (string)": 95.259,
  "sort (multi-col)": 216.636,
  "mutate (col/scalar)": 1.969,
  "mutate (col+col)": 2.127,
  "mutate (string upper)": 39.760,
  "mutate (scalar)": 0.013,
  "distinct": 87.022,
  "groupBy (single)": 3.982,
  "groupBy (multi)": 11.251,
  "summarise (ungrouped)": 0.451,
  "summarise (grouped)": 19.398,
  "innerJoin": 54.833,
  "leftJoin": 68.612,
  "pivotLonger": 45.331,
  "bindRows": 0.036,
  "stats": 8.503,
};

const PANDAS: Record<string, number> = {
  "creation": 126.866,
  "filter (numeric)": 15.518,
  "filter (string)": 10.548,
  "filter (complex)": 7.514,
  "select": 0.189,
  "sort (numeric)": 301.104,
  "sort (string)": 470.756,
  "sort (multi-col)": 854.214,
  "mutate (col/scalar)": 0.835,
  "mutate (col+col)": 0.959,
  "mutate (string upper)": 21.248,
  "mutate (scalar)": 0.695,
  "distinct": 419.233,
  "groupBy (single)": 23.618,
  "groupBy (multi)": 37.184,
  "summarise (ungrouped)": 5.097,
  "summarise (grouped)": 33.479,
  "innerJoin": 278.515,
  "leftJoin": 276.623,
  "pivotLonger": 41.450,
  "bindRows": 1.281,
  "stats": 38.808,
};

function median(arr: number[]): number {
  const sorted = [...arr].sort((a, b) => a - b);
  return sorted[Math.floor(sorted.length / 2)];
}

function bench(label: string, fn: () => unknown): number {
  for (let i = 0; i < WARMUP; i++) fn();
  const times: number[] = [];
  for (let i = 0; i < ITERS; i++) {
    const t = performance.now();
    fn();
    times.push(performance.now() - t);
  }
  return median(times);
}

console.log(`\n=== Broad Benchmark (${(N/1e6).toFixed(0)}M rows, ${ITERS} iters, ${WARMUP} warmup) ===\n`);

// Generate data
console.log("Generating data...");
const t0 = performance.now();
const data = Array.from({ length: N }, (_, i) => ({
  id: i + 1,
  value: Math.random() * 1000,
  category: `cat_${i % 20}`,
  score: Math.random() * 100,
  active: i % 3 === 0,
}));
console.log(`  Data generated: ${(performance.now() - t0).toFixed(0)}ms`);

// Prebuild
console.log("Building DataFrame...");
const t1 = performance.now();
const df = createDataFrame(data);
console.log(`  DataFrame built: ${(performance.now() - t1).toFixed(0)}ms\n`);

// Join data
const joinRight = createDataFrame(
  Array.from({ length: Math.floor(N * 0.8) }, (_, i) => ({
    id: Math.floor(Math.random() * N) + 1,
    value_b: Math.random() * 1000,
    status: ["active", "pending", "complete"][i % 3],
  }))
);

// Split data for bindRows
const half = Math.floor(N / 2);
const df1 = createDataFrame(data.slice(0, half));
const df2 = createDataFrame(data.slice(half));

// Pivot data
const pivotData = createDataFrame(
  Array.from({ length: N }, (_, i) => ({
    id: i + 1,
    region: `region_${i % 5}`,
    product: `product_${i % 10}`,
    q1: Math.floor(Math.random() * 1000),
    q2: Math.floor(Math.random() * 1000),
    q3: Math.floor(Math.random() * 1000),
    q4: Math.floor(Math.random() * 1000),
  }))
);

const results: { op: string; ms: number }[] = [];

function run(label: string, fn: () => unknown) {
  const ms = bench(label, fn);
  results.push({ op: label, ms });
  console.log(`  ${label.padEnd(30)} ${ms.toFixed(3)}ms`);
}

// 1. Creation
run("creation", () => createDataFrame(data));

// 2. Filter (numeric)
run("filter (numeric)", () => df.filter((r) => r.value > 500));

// 3. Filter (string)
run("filter (string)", () => df.filter((r) => r.category === "cat_5"));

// 4. Filter (complex)
run("filter (complex)", () => df.filter((r) => r.value > 300 && r.score > 50 && r.active));

// 5. Select
run("select", () => df.select("id", "value", "category"));

// 6. Sort (numeric)
run("sort (numeric)", () => df.arrange("value", "asc"));

// 7. Sort (string)
run("sort (string)", () => df.arrange("category", "asc"));

// 8. Sort (multi-col)
run("sort (multi-col)", () => df.arrange(["category", "value"], ["asc", "desc"]));

// 9. Mutate (col/scalar)
run("mutate (col/scalar)", () => df.mutate({ score_pct: (r) => r.score / 100 }));

// 10. Mutate (col+col)
run("mutate (col+col)", () => df.mutate({ total: (r) => r.value + r.score }));

// 11. Mutate (string)
run("mutate (string upper)", () => df.mutate({ cat_upper: (r) => r.category.toUpperCase() }));

// 12. Mutate (scalar)
run("mutate (scalar)", () => df.mutate({ constant: 42 }));

// 13. Distinct
run("distinct", () => df.distinct("id", "value", "category"));

// 14. GroupBy (single)
run("groupBy (single)", () => df.groupBy("category"));

// 15. GroupBy (multi)
run("groupBy (multi)", () => df.groupBy("category", "active"));

// 16. Summarise (ungrouped)
run("summarise (ungrouped)", () =>
  df.summarise({
    count: (g) => g.nrows(),
    avg_value: (g) => stats.mean(g.value),
    total_value: (g) => stats.sum(g.value),
  })
);

// 17. Summarise (grouped)
run("summarise (grouped)", () =>
  df.groupBy("category").summarise({
    count: (g) => g.nrows(),
    avg_value: (g) => stats.mean(g.value),
    total_value: (g) => stats.sum(g.value),
  })
);

// 18. Inner Join
run("innerJoin", () => df.select("id", "value").innerJoin(joinRight, "id"));

// 19. Left Join
run("leftJoin", () => df.select("id", "value").leftJoin(joinRight, "id"));

// 20. Pivot Longer
run("pivotLonger", () =>
  pivotData.pivotLonger({
    cols: ["q1", "q2", "q3", "q4"],
    namesTo: "quarter",
    valuesTo: "sales",
  })
);

// 21. Bind Rows
run("bindRows", () => df1.bindRows(df2));

// 22. Stats
run("stats", () => {
  const values = df.value as number[];
  stats.sum(values);
  stats.mean(values);
  stats.median(values);
  stats.variance(values);
  stats.stdev(values);
});

// ── 3-way comparison table ──
const pad = (s: string, n: number) => s.padEnd(n);
const rpad = (s: string, n: number) => s.padStart(n);

// Collect rows with delta for sorting
const rows: { op: string; ms: number; pl: number; pd: number; deltaPolars: number; deltaPandas: number }[] = [];
let totalTidy = 0, totalPolars = 0, totalPandas = 0;

for (const { op, ms } of results) {
  const pl = POLARS[op] ?? 0;
  const pdMs = PANDAS[op] ?? 0;
  totalTidy += ms;
  totalPolars += pl;
  totalPandas += pdMs;
  rows.push({ op, ms, pl, pd: pdMs, deltaPolars: ms - pl, deltaPandas: ms - pdMs });
}

// Sort by Δ vs Polars descending — biggest time gaps first
const sorted = [...rows].sort((a, b) => b.deltaPolars - a.deltaPolars);

const W = 100;
console.log(`\n${"─".repeat(W)}`);
console.log(`  ${pad("Operation", 30)} ${rpad("tidy-ts", 10)} ${rpad("Polars", 10)} ${rpad("pandas", 10)} ${rpad("Δ Polars", 10)} ${rpad("Δ pandas", 10)} ${rpad("vs Polars", 10)} ${rpad("vs pandas", 10)}`);
console.log(`${"─".repeat(W)}`);

for (const { op, ms, pl, pd, deltaPolars, deltaPandas } of sorted) {
  const fmtDelta = (d: number) => {
    const sign = d >= 0 ? "+" : "";
    return `${sign}${d.toFixed(1)}ms`;
  };
  const vsPolars = pl > 0 ? `${(ms / pl).toFixed(1)}x` : "—";
  const vsPandas = pd > 0 ? `${(ms / pd).toFixed(1)}x` : "—";

  console.log(
    `  ${pad(op, 30)} ${rpad(ms.toFixed(1) + "ms", 10)} ${rpad(pl.toFixed(1) + "ms", 10)} ${rpad(pd.toFixed(1) + "ms", 10)} ${rpad(fmtDelta(deltaPolars), 10)} ${rpad(fmtDelta(deltaPandas), 10)} ${rpad(vsPolars, 10)} ${rpad(vsPandas, 10)}`
  );
}

console.log(`${"─".repeat(W)}`);
const fmtTotalDelta = (d: number) => `${d >= 0 ? "+" : ""}${d.toFixed(1)}ms`;
console.log(
  `  ${pad("TOTAL", 30)} ${rpad(totalTidy.toFixed(1) + "ms", 10)} ${rpad(totalPolars.toFixed(1) + "ms", 10)} ${rpad(totalPandas.toFixed(1) + "ms", 10)} ${rpad(fmtTotalDelta(totalTidy - totalPolars), 10)} ${rpad(fmtTotalDelta(totalTidy - totalPandas), 10)} ${rpad((totalTidy / totalPolars).toFixed(1) + "x", 10)} ${rpad((totalTidy / totalPandas).toFixed(1) + "x", 10)}`
);

console.log("\nDone.");

File: packages/testing/benchmarks/bench-broad-2m.py
#!/usr/bin/env python3
"""
Broad Python benchmark at 2M rows (pandas + Polars).
Matches bench-broad-2m.ts operations.
"""
import polars as pl
import pandas as pd
import numpy as np
import time
import os

os.environ["POLARS_MAX_THREADS"] = "4"

N = 2_000_000
ITERS = 10
WARMUP = 5

def median(arr):
    s = sorted(arr)
    return s[len(s) // 2]

def bench(label, fn):
    for _ in range(WARMUP):
        fn()
    times = []
    for _ in range(ITERS):
        t0 = time.perf_counter()
        fn()
        times.append((time.perf_counter() - t0) * 1000)
    return median(times)

print(f"\n=== Broad Python Benchmark ({N//1_000_000}M rows, {ITERS} iters, {WARMUP} warmup) ===\n")

# Generate raw data
print("Generating data...")
rng = np.random.default_rng(42)
ids = np.arange(1, N + 1)
values = rng.random(N) * 1000
categories = [f"cat_{i % 20}" for i in range(N)]
scores = rng.random(N) * 100
actives = [i % 3 == 0 for i in range(N)]

data_dict = {
    "id": ids,
    "value": values,
    "category": categories,
    "score": scores,
    "active": actives,
}

# Build DataFrames
pl_df = pl.DataFrame(data_dict)
pd_df = pd.DataFrame(data_dict)
print("  DataFrames built\n")

# Join data
right_size = int(N * 0.8)
join_ids = rng.integers(1, N + 1, right_size)
join_values_b = rng.random(right_size) * 1000
join_status = [["active", "pending", "complete"][i % 3] for i in range(right_size)]

pl_join_right = pl.DataFrame({"id": join_ids, "value_b": join_values_b, "status": join_status})
pd_join_right = pd.DataFrame({"id": join_ids, "value_b": join_values_b, "status": join_status})

# Split for bind
pl_df1 = pl_df.head(N // 2)
pl_df2 = pl_df.tail(N - N // 2)
pd_df1 = pd_df.head(N // 2)
pd_df2 = pd_df.tail(N - N // 2)

# Pivot data
pivot_ids = np.arange(1, N + 1)
pivot_regions = [f"region_{i % 5}" for i in range(N)]
pivot_products = [f"product_{i % 10}" for i in range(N)]
pivot_q1 = rng.integers(0, 1000, N)
pivot_q2 = rng.integers(0, 1000, N)
pivot_q3 = rng.integers(0, 1000, N)
pivot_q4 = rng.integers(0, 1000, N)

pl_pivot = pl.DataFrame({
    "id": pivot_ids, "region": pivot_regions, "product": pivot_products,
    "q1": pivot_q1, "q2": pivot_q2, "q3": pivot_q3, "q4": pivot_q4,
})
pd_pivot = pd.DataFrame({
    "id": pivot_ids, "region": pivot_regions, "product": pivot_products,
    "q1": pivot_q1, "q2": pivot_q2, "q3": pivot_q3, "q4": pivot_q4,
})

# Pre-select for joins
pl_id_val = pl_df.select(["id", "value"])
pd_id_val = pd_df[["id", "value"]]

# ── Results storage ──
polars_results = []
pandas_results = []

def run_both(label, polars_fn, pandas_fn):
    pl_ms = bench(label, polars_fn)
    pd_ms = bench(label, pandas_fn)
    polars_results.append((label, pl_ms))
    pandas_results.append((label, pd_ms))
    print(f"  {label:<30} polars={pl_ms:>9.3f}ms   pandas={pd_ms:>9.3f}ms   ratio={pd_ms/pl_ms:>6.1f}x")

# 1. Creation
run_both("creation",
    lambda: pl.DataFrame(data_dict),
    lambda: pd.DataFrame(data_dict),
)

# 2. Filter (numeric)
run_both("filter (numeric)",
    lambda: pl_df.filter(pl.col("value") > 500),
    lambda: pd_df[pd_df["value"] > 500],
)

# 3. Filter (string)
run_both("filter (string)",
    lambda: pl_df.filter(pl.col("category") == "cat_5"),
    lambda: pd_df[pd_df["category"] == "cat_5"],
)

# 4. Filter (complex)
run_both("filter (complex)",
    lambda: pl_df.filter((pl.col("value") > 300) & (pl.col("score") > 50) & pl.col("active")),
    lambda: pd_df[(pd_df["value"] > 300) & (pd_df["score"] > 50) & pd_df["active"]],
)

# 5. Select
run_both("select",
    lambda: pl_df.select(["id", "value", "category"]),
    lambda: pd_df[["id", "value", "category"]],
)

# 6. Sort (numeric)
run_both("sort (numeric)",
    lambda: pl_df.sort("value"),
    lambda: pd_df.sort_values("value"),
)

# 7. Sort (string)
run_both("sort (string)",
    lambda: pl_df.sort("category"),
    lambda: pd_df.sort_values("category", kind="mergesort"),
)

# 8. Sort (multi-col)
run_both("sort (multi-col)",
    lambda: pl_df.sort(["category", "value"], descending=[False, True]),
    lambda: pd_df.sort_values(["category", "value"], ascending=[True, False]),
)

# 9. Mutate (col/scalar)
run_both("mutate (col/scalar)",
    lambda: pl_df.with_columns((pl.col("score") / 100).alias("score_pct")),
    lambda: pd_df.assign(score_pct=pd_df["score"] / 100),
)

# 10. Mutate (col+col)
run_both("mutate (col+col)",
    lambda: pl_df.with_columns((pl.col("value") + pl.col("score")).alias("total")),
    lambda: pd_df.assign(total=pd_df["value"] + pd_df["score"]),
)

# 11. Mutate (string upper)
run_both("mutate (string upper)",
    lambda: pl_df.with_columns(pl.col("category").str.to_uppercase().alias("cat_upper")),
    lambda: pd_df.assign(cat_upper=pd_df["category"].str.upper()),
)

# 12. Mutate (scalar)
run_both("mutate (scalar)",
    lambda: pl_df.with_columns(pl.lit(42).alias("constant")),
    lambda: pd_df.assign(constant=42),
)

# 13. Distinct
run_both("distinct",
    lambda: pl_df.unique(),
    lambda: pd_df.drop_duplicates(),
)

# 14. GroupBy (single)
run_both("groupBy (single)",
    lambda: pl_df.group_by("category").len(),
    lambda: pd_df.groupby("category", sort=False).size().reset_index(name="len"),
)

# 15. GroupBy (multi)
run_both("groupBy (multi)",
    lambda: pl_df.group_by(["category", "active"]).len(),
    lambda: pd_df.groupby(["category", "active"], sort=False).size().reset_index(name="len"),
)

# 16. Summarise (ungrouped)
run_both("summarise (ungrouped)",
    lambda: pl_df.select([
        pl.count("id").alias("count"),
        pl.mean("value").alias("avg_value"),
        pl.sum("value").alias("total_value"),
    ]),
    lambda: pd.DataFrame({
        "count": [pd_df["id"].count()],
        "avg_value": [pd_df["value"].mean()],
        "total_value": [pd_df["value"].sum()],
    }),
)

# 17. Summarise (grouped)
run_both("summarise (grouped)",
    lambda: pl_df.group_by("category").agg([
        pl.count("id").alias("count"),
        pl.mean("value").alias("avg_value"),
        pl.sum("value").alias("total_value"),
    ]),
    lambda: pd_df.groupby("category", sort=False).agg(
        count=("id", "count"),
        avg_value=("value", "mean"),
        total_value=("value", "sum"),
    ).reset_index(),
)

# 18. Inner Join
run_both("innerJoin",
    lambda: pl_id_val.join(pl_join_right, on="id", how="inner"),
    lambda: pd_id_val.merge(pd_join_right, on="id", how="inner"),
)

# 19. Left Join
run_both("leftJoin",
    lambda: pl_id_val.join(pl_join_right, on="id", how="left"),
    lambda: pd_id_val.merge(pd_join_right, on="id", how="left"),
)

# 20. Pivot Longer
run_both("pivotLonger",
    lambda: pl_pivot.unpivot(
        index=["id", "region", "product"],
        on=["q1", "q2", "q3", "q4"],
        variable_name="quarter",
        value_name="sales",
    ),
    lambda: pd_pivot.melt(
        id_vars=["id", "region", "product"],
        value_vars=["q1", "q2", "q3", "q4"],
        var_name="quarter",
        value_name="sales",
    ),
)

# 21. Bind Rows
run_both("bindRows",
    lambda: pl.concat([pl_df1, pl_df2]),
    lambda: pd.concat([pd_df1, pd_df2], ignore_index=True),
)

# 22. Stats
run_both("stats",
    lambda: pl_df.select([
        pl.col("value").sum().alias("sum"),
        pl.col("value").mean().alias("mean"),
        pl.col("value").median().alias("median"),
        pl.col("value").var().alias("variance"),
        pl.col("value").std().alias("stdev"),
    ]),
    lambda: (
        pd_df["value"].sum(),
        pd_df["value"].mean(),
        pd_df["value"].median(),
        pd_df["value"].var(),
        pd_df["value"].std(),
    ),
)

# ── Summary table ──
print(f"\n{'─'*75}")
print(f"  {'Operation':<30} {'Polars':>9}   {'pandas':>9}   {'pd/pl':>6}")
print(f"{'─'*75}")
total_pl = 0
total_pd = 0
for (label, pl_ms), (_, pd_ms) in zip(polars_results, pandas_results):
    total_pl += pl_ms
    total_pd += pd_ms
    print(f"  {label:<30} {pl_ms:>8.3f}ms   {pd_ms:>8.3f}ms   {pd_ms/pl_ms:>5.1f}x")
print(f"{'─'*75}")
print(f"  {'TOTAL':<30} {total_pl:>8.3f}ms   {total_pd:>8.3f}ms   {total_pd/total_pl:>5.1f}x")

print("\nDone.")
